### ISSUE TITLE
Bool query with must query is matching all documents when the query doesn't match

### ISSUE THREAD

**Issue by rayward** [elastic/elasticsearch#9866]
We're seeing an issue where if a query, such as a `simple_query_string` used in `must` clause (or multiple queries in `should` clauses) of a `bool` query, and the analyzed query string results in an empty query string (such as stop words being filtered out), then the it behaves like the entire query is ignored and all documents are matched. Even if the `simple_query_string` query is inside a `filtered` query, then the filters are ignored.

This is causing major issues for us in a multi-tenant environment as we're getting results not belonging to the correct tenant.

This issue started after we recently upgraded from 1.2.1 to 1.4.3.

This is issue appears to have been introduced in 1.3.3 (from testing various versions) and also affects 1.4.4.

I've simplified our queries below with only query in the `must` clause. Real world, we're searching against multiple types.

Analysis:

```json
curl -XPUT 'localhost:9200/my_index' -d '
{
  "index": {
    "analysis": {
      "analyzer": {
        "html_strip": {
          "filter": [
            "standard",
            "lowercase",
            "stop",
            "asciifolding",
            "minimal_stemmer"
          ],
          "char_filter": [
            "html_strip"
          ],
          "tokenizer": "standard",
          "type": "custom"
        },
        "default_search": {
          "filter": [
            "standard",
            "lowercase",
            "stop",
            "asciifolding",
            "minimal_stemmer"
          ],
          "tokenizer": "standard",
          "type": "custom"
        },
        "default_index": {
          "filter": [
            "standard",
            "lowercase",
            "stop",
            "asciifolding",
            "minimal_stemmer"
          ],
          "tokenizer": "standard",
          "type": "custom"
        }
      },
      "filter": {
        "minimal_stemmer": {
          "type": "stemmer",
          "name": "minimal_english"
        }
      }
    }
  }
}
'
```

Create `page` type:
```json
curl -XPUT 'localhost:9200/my_index/page/_mapping' -d '
{
  "page" : {
    "properties" : {
      "title" : {
        "type" : "string"
      },
      "content" : {
        "type" : "string",
        "analyzer": "html_strip"
      },
      "is_visible" : {
        "type" : "boolean"
      },
      "store_id": {
        "index": "not_analyzed",
        "type": "integer"
      }
    }
  }
}
'
```

Create two documents with different `store_id`'s:

```json
curl -XPOST 'localhost:9200/my_index/page/store-1234-id-1' -d '
{
  "title": "Page for store 1234",
  "content": "The content for this page",
  "is_visible": true,
  "store_id": 1234
}
'

curl -XPOST 'localhost:9200/my_index/page/store-5678-id-1' -d '
{
  "title": "Page for store 5678",
  "content": "The content for another page",
  "is_visible": true,
  "store_id": 5678
}
'
```

Execute a search containing just a stop word `the`:

```json
curl -XPOST 'localhost:9200/my_index/page/_search' -d '
{
   "explain": true,
    "query": {
        "bool": {
            "must": [
                {
                    "filtered": {
                        "query": {
                            "simple_query_string": {
                                "query": "the",
                                "fields": [
                                    "title",
                                    "content"
                                ]
                            }
                        },
                        "filter": {
                            "and": [
                                {
                                    "term": {
                                        "is_visible": true
                                    }
                                },
                                {
                                    "term": {
                                        "store_id": 1234
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }
    }
}'
```

Result:

```json
{
  "took": 3,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "failed": 0
  },
  "hits": {
    "total": 2,
    "max_score": 1,
    "hits": [
      {
        "_shard": 1,
        "_node": "HGgHSCGZTwqLOWwDwSEo6w",
        "_index": "my_index",
        "_type": "page",
        "_id": "store-1234-id-1",
        "_score": 1,
        "_source": {
          "title": "Page for store 1234",
          "content": "The content for this page",
          "is_visible": true,
          "store_id": 1234
        },
        "_explanation": {
          "value": 1,
          "description": "ConstantScore(cache(_type:page)), product of:",
          "details": [
            {
              "value": 1,
              "description": "boost"
            },
            {
              "value": 1,
              "description": "queryNorm"
            }
          ]
        }
      },
      {
        "_shard": 4,
        "_node": "HGgHSCGZTwqLOWwDwSEo6w",
        "_index": "my_index",
        "_type": "page",
        "_id": "store-5678-id-1",
        "_score": 1,
        "_source": {
          "title": "Page for store 5678",
          "content": "The content for another page",
          "is_visible": true,
          "store_id": 5678
        },
        "_explanation": {
          "value": 1,
          "description": "ConstantScore(cache(_type:page)), product of:",
          "details": [
            {
              "value": 1,
              "description": "boost"
            },
            {
              "value": 1,
              "description": "queryNorm"
            }
          ]
        }
      }
    ]
  }
}
```

As you can see both documents are returned.

When the query contains a different search term that _does_ exist in a document (eg `content`):

```json
curl -XPOST 'localhost:9200/my_index/page/_search' -d '{
    "explain": true,
    "query": {
        "bool": {
            "must": [
                {
                    "filtered": {
                        "query": {
                            "simple_query_string": {
                                "query": "content",
                                "fields": [
                                    "name",
                                    "content"
                                ]
                            }
                        },
                        "filter": {
                            "and": [
                                {
                                    "term": {
                                        "is_visible": true
                                    }
                                },
                                {
                                    "term": {
                                        "store_id": 1234
                                    }
                                }
                            ]
                        }
                    }
                }
            ]
        }
    }
}
'
```

Result:

```json
{
  "took": 4,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "failed": 0
  },
  "hits": {
    "total": 1,
    "max_score": 0.05626005,
    "hits": [
      {
        "_shard": 1,
        "_node": "HGgHSCGZTwqLOWwDwSEo6w",
        "_index": "my_index",
        "_type": "page",
        "_id": "store-1234-id-1",
        "_score": 0.05626005,
        "_source": {
          "title": "Page for store 1234",
          "content": "The content for this page",
          "is_visible": true,
          "store_id": 1234
        },
        "_explanation": {
          "value": 0.056260053,
          "description": "sum of:",
          "details": [
            {
              "value": 0.056260053,
              "description": "weight(content:content in 0) [PerFieldSimilarity], result of:",
              "details": [
                {
                  "value": 0.056260053,
                  "description": "score(doc=0,freq=1.0), product of:",
                  "details": [
                    {
                      "value": 0.29335263,
                      "description": "queryWeight, product of:",
                      "details": [
                        {
                          "value": 0.30685282,
                          "description": "idf(docFreq=1, maxDocs=1)"
                        },
                        {
                          "value": 0.9560043,
                          "description": "queryNorm"
                        }
                      ]
                    },
                    {
                      "value": 0.19178301,
                      "description": "fieldWeight in 0, product of:",
                      "details": [
                        {
                          "value": 1,
                          "description": "tf(freq=1.0), with freq of:",
                          "details": [
                            {
                              "value": 1,
                              "description": "termFreq=1.0"
                            }
                          ]
                        },
                        {
                          "value": 0.30685282,
                          "description": "idf(docFreq=1, maxDocs=1)"
                        },
                        {
                          "value": 0.625,
                          "description": "fieldNorm(doc=0)"
                        }
                      ]
                    }
                  ]
                }
              ]
            }
          ]
        }
      }
    ]
  }
}
```

When the query is executed outside the context of a `bool` query:

```json
curl -XPOST 'localhost:9200/my_index/page/_search' -d '{
    "explain": true,
    "query": {
        "filtered": {
            "query": {
                "simple_query_string": {
                    "query": "the",
                    "fields": [
                        "name",
                        "content"
                    ]
                }
            },
            "filter": {
                "and": [
                    {
                        "term": {
                            "is_visible": true
                        }
                    },
                    {
                        "term": {
                            "store_id": 1234
                        }
                    }
                ]
            }
        }
    }
}
'
```

We now correctly get no results:

```json
{
  "took": 1,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "failed": 0
  },
  "hits": {
    "total": 0,
    "max_score": null,
    "hits": []
  }
}
```

In 1.3.2 or below you'll get the above empty result when using the original bool query.

Has anything changed in mappings or analysis that's causing this or is this a regression?

**Comment by rayward** [comment]
Interestingly, if I change `simple_query_string` to `query_string`, then I get the correct results.

Is this a combination of `simple_query_string` not strictly having this behaviour from `query_string`:

>empty query
> If the query string is empty or only contains whitespaces the query will yield an empty result set.

And this change in 1.3.3 - https://github.com/elasticsearch/elasticsearch/pull/7347 ?

**Comment by rayward** [comment]
We've confirmed that reverting this particular change - https://github.com/elasticsearch/elasticsearch/pull/7347/files#diff-94095d9e5782634b609ae9a773f11827L134 - resolves the issue for us.

I don't believe that is the "correct" fix in the end though, but it does restore the previous behaviour regarding `simple_query_string`.

**Comment by dakrone** [comment]
Sorry, I didn't mean to close this, I am looking into this :)

**Comment by dakrone** [comment]
As part of the root cause, I've submitted a patch for this to Lucene here: https://issues.apache.org/jira/browse/LUCENE-6298

**Comment by rayward** [comment]
Thanks @dakrone :+1:

**Comment by clintongormley** [comment]
Closed by #10435
